File: 백준/Gold/17182. 우주 탐사선/우주 탐사선.py
import sys
input=sys.stdin.readline

def find(node,cnt,visit):
  global answer
  if len(visit)==N:
    if answer==-1 or answer>cnt:
      answer=cnt
    return

  for i in range(N):
    if i==node or i in visit:
      continue
    newVisit=list(visit)
    newVisit.append(i)
    find(i,cnt+l[node][i],newVisit)

N,K=map(int,input().split())
l=[]
for i in range(N):
  l.append(list(map(int,input().split())))

for k in range(N):
  for i in range(N):
    for j in range(N):
      l[i][j] = min(l[i][j],l[i][k]+l[k][j])

answer=-1
find(K,0,[K])
print(answer)

File: 백준/Gold/17182. 우주 탐사선/우주 탐사선.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let input = [];
rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  [N, start] = input.shift().split(' ').map(Number);

  function find(now, cnt) {
    if (cnt >= answer) return;
    if (now.length === N) {
      answer = Math.min(answer, cnt);
      return;
    }
    for (let i = 0; i < N; i++) {
      if (now.includes(i)) continue;
      now.push(i);
      find(now, cnt + visited[now[now.length - 2]][i]);
      now.pop();
    }
  }

  for (let i = 0; i < N; i++) {
    input[i] = input[i].split(' ').map(Number);
  }

  let visited = [];
  for (let i = 0; i < N; i++) {
    visited.push([]);
    visited[i] = input[i];
  }

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (j == i) continue;
      for (let k = 0; k < N; k++) {
        visited[i][j] = Math.min(visited[i][j], visited[i][k] + visited[k][j]);
      }
    }
  }

  let answer = 20000;
  find([start], 0);
  console.log(answer);
  process.exit();
});
